wcfsetup/install/files/lib/system/search/SearchHandler.class.php

```diff
@@ -0,0 +1,347 @@
+    {
+        foreach ($this->parameters as $key => $value) {
+            $_POST[$key] = $value;
+        }
+
+        $form = new SearchForm();
+        $form->readFormParameters();
+        $form->userIDs = $this->getUserIDs();
+        if (\count($form->selectedObjectTypes) === 1) {
+            $this->objectTypeNames = $form->selectedObjectTypes;
+        }
+        if ($form->sortField) {
+            $this->parameters['sortField'] = $form->sortField;
+        }
+        if ($form->sortOrder) {
+            $this->parameters['sortOrder'] = $form->sortOrder;
+        }
+
+        return $form;
+    }
+
+    private function getUserIDs(): array
+    {
+        if ($this->userIDs === null) {
+            $this->userIDs = [];
+
+            if (!empty($this->parameters['username'])) {
+                $sql = "SELECT  userID
+                        FROM    wcf" . WCF_N . "_user
+                        WHERE   username " . (!empty($this->parameters['nameExactly']) ? "= ?" : "LIKE ?");
+                $statement = WCF::getDB()->prepareStatement($sql, 100);
+                $statement->execute([$this->parameters['username'] . (empty($this->parameters['nameExactly']) ? '%' : '')]);
+                $this->userIDs = $statement->fetchAll(\PDO::FETCH_COLUMN);
+            }
+
+            if (!empty($this->parameters['userID'])) {
+                $this->userIDs[] = $this->parameters['userID'];
+            }
+        }
+
+        return $this->userIDs;
+    }
+
+    private function saveKeywordSuggestion(): void
+    {
+        if (!empty($this->parameters['q'])) {
+            (new SearchKeywordAction([], 'registerSearch', [
+                'data' => [
+                    'keyword' => $this->parameters['q'],
+                ],
+            ]))->executeAction();
+        }
+    }
+
+    private function buildSearchHash(): void
+    {
+        $this->searchHash = \sha1(\serialize([
+            $this->parameters['q'] ?? '',
+            $this->objectTypeNames,
+            $this->parameters['subjectOnly'] ?? 0,
+            $this->conditionBuilder,
+            $this->typeBasedConditionBuilders,
+            $this->parameters['sortField'],
+            $this->parameters['sortOrder'],
+            $this->getAdditionalData(),
+        ]));
+    }
+
+    private function getExistingSearch(): ?Search
+    {
+        if (!empty($this->parameters['q'])) {
+            $parameters = [$this->searchHash, 'messages', TIME_NOW - 1800];
+            if (WCF::getUser()->userID) {
+                $parameters[] = WCF::getUser()->userID;
+            }
+
+            $sql = "SELECT  searchID
+                    FROM    wcf" . WCF_N . "_search
+                    WHERE   searchHash = ?
+                        AND searchType = ?
+                        AND searchTime > ?
+                        " . (WCF::getUser()->userID ? 'AND userID = ?' : 'AND userID IS NULL');
+            $statement = WCF::getDB()->prepareStatement($sql);
```

Comment:
The condition handling could be cleaned up with a PreparedStatementConditionBuilder.